build.gradle.kts

```diff
@@ -56,6 +58,57 @@ subprojects {
     }
 
     tasks.register<DependencyReportTask>("allDependencies") {}
+
+    tasks.register("applyDependencyRules") {
+        doLast {
+
+            fun dependencyError(error: String) {
+                val message = "DEPENDENCY RULE VIOLATION: $error"
+                if (dependencyAnalysis == "fail") {
+                    throw GradleException(message)
+                } else {
+                    println(message)
+                }
+            }
+
+            val dependencies = configurations["compileClasspath"].resolvedConfiguration.resolvedArtifacts
+
+            dependencies.forEach { artifact ->
+                val dependency = artifact.moduleVersion.id
+
+                if (dependency.group == project.group) {
+                    val pathFromRoot = artifact.file.relativeTo(project.rootDir).path
+                    val pathFromThisModule = artifact.file.relativeTo(project.projectDir).path
+
+                    if (!dependency.name.endsWith("-spi") // modules may only depend on `-spi` modules (exceptions follow)
+                        && !dependency.name.endsWith("spi")
+                        && !dependency.name.endsWith("-core") // exception: technology libs such as "blob-core"
+                        && !pathFromRoot.startsWith("common/") // exception: `common`
+                        && !pathFromRoot.startsWith("extensions/http/jetty/") // exception: `jetty` (this exception should be removed once there is an SPI for jetty)
+                        && !project.path.startsWith(":launchers:") // exception: launchers
+                        && !project.path.startsWith(":samples:") // exception: samples
+                        && !project.path.startsWith(":system-tests:") // exception: system-tests
+                    ) {
+                        dependencyError("modules may only depend on '*-spi' modules. Invalid dependency: $dependency")
+                    }
+
+                    if (pathFromThisModule.matches(Regex("\\.\\./[^.].*")) // there should not be "cross-module" dependencies at the same level
+                        && !dependency.name.endsWith("-core") // exception: technology libs such as "blob-core"
+                    ) {
+                        dependencyError("there should not be \"cross-module\" dependencies at the same level. Invalid dependency: $dependency")
+                    }
+
+                    if (project.name == "core-spi") { // `core:spi` cannot depend on any other module
+                        dependencyError("`core:spi` cannot depend on any other module. Invalid dependency: $dependency")
+                    }
+
+                    if (project.name == dependency.name) { // two modules cannot have the same name (TBC)
+                        dependencyError("two modules cannot have the same name. Invalid dependency: $dependency")
+                    }
+                }
```

Comment:
I think we are missing rules for the following cases:
- no module may depend directly on any of the core modules
- no module may depend on a launcher, samples or system-tests module

Reply:
Right, I didn't have a way to reason on the *path* to a dependency initially, though now there is one with `pathFromRoot` and these can be implemented

Reply:
Ok I can take over this

Reply:
Done.